### PR TITLE
Decouple PairingState from PeerConnectionState

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
PASESession and CASESession are reusing `PeerConnectionState` to store pairing states, but actually they are different things.

I'm going to make `SecureSessionManager` also handle unsecure message, where `PeerConnectionState` must be extended to handle unsecure connections. So I need to decouple `PairingState` from `PeerConnectionState`.

#### Change overview
Create a new class `PairingState` and switch `mConnection` in  PASESession and CASESession to that class.

#### Testing
Verified using unit-tests

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

File: src/protocols/secure_channel/PairingState.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include <transport/SecureSession.h>
+#include <transport/SessionMessageCounter.h>
+#include <transport/raw/Base.h>
+#include <transport/raw/MessageHeader.h>
+#include <transport/raw/PeerAddress.h>
+
+namespace chip {
+
+/**
+ * Defines state of a pairing session
```
Comment (andy31415) [inline src/protocols/secure_channel/PairingState.h:30]: Comment feels obvious: `PairingState` is obviously "a state of pairing".

Could we update the documentation in some way to add more interesting stuff? Like what is a pairing state, and how  is it supposed to be used.

File: src/protocols/secure_channel/PairingState.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        mPeerAddress = Transport::PeerAddress::Uninitialized();
+        mPeerNodeId  = kUndefinedNodeId;
+        mPeerKeyID   = UINT16_MAX;
+        mLocalKeyID  = UINT16_MAX;
+    }
+
+private:
+    Transport::PeerAddress mPeerAddress = Transport::PeerAddress::Uninitialized();
+    NodeId mPeerNodeId                  = kUndefinedNodeId;
+    uint16_t mPeerKeyID                 = UINT16_MAX;
```
Comment (andy31415) [inline src/protocols/secure_channel/PairingState.h:68]: please add a class `static constexpr uint16_t kInvalidKeyId = UINT16_MAX` and use that.

This way one can clearly compare with `PairingState::kInvalidKeyId` for keys or we could add a 'bool IsValidPeerKeyId() const;` and similar

File: src/protocols/secure_channel/PairingState.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#pragma once
+
+#include <transport/AdminPairingTable.h>
+#include <transport/SecureSession.h>
+#include <transport/SessionMessageCounter.h>
+#include <transport/raw/Base.h>
+#include <transport/raw/MessageHeader.h>
+#include <transport/raw/PeerAddress.h>
+
+namespace chip {
```
Comment (andy31415) [inline src/protocols/secure_channel/PairingState.h:27]: Is top-level chip namespace ok or should we go deeper and say 'SecureChannel' or similar?

PeerConnectionState is in 'transport'

**Review by andy31415 — APPROVED** [review]

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
What is the plan for how PairingState will diverge from PeerConnectionState in the future?  Should they have some common superclass, perhaps?

Marking changes requested until @pan-apple has a chance to take a look, but also because of very much insufficient documentation (even less than PeerConnectionState has), and the lack of clarity about how much of this functionality duplication is really needed.

**Comment by kghost** [comment]
@bzbarsky-apple @pan-apple #8206 This is the direction I'm working on.

**Comment by bzbarsky-apple** [comment]
OK.  But then I don't quite understand why the pairing state can't point to a PeerCacheEntry....

**Comment by kghost** [comment]
The cache entry is accessible via `ExchangeContext` so there is no need to duplicate it again.

**Comment by pan-apple** [comment]
@kghost it seems the new `PairingState` object is relevant only for CASE and PASE handshakes. If so, we can integrate this state inside `PairingSession` instead of creating a new class. Both, `PASESession` and `CASESession`, derive from `PairingSession` already. So they'll have access to the state.

**Comment by bzbarsky-apple** [comment]
@kghost I think the thing that confuses/bothers me is the multiple of classes with a bunch of the same shared information in them which are mostly copy/pastes of each other.  It suggests that we need to figure out some better representation for these thing (e.g. that we should have some concept that is shared by all these classes and that we might need to come up with a name for)....

So let's try to do that:

1. What does PeerConnectionState represent?
2. What does PairingState represent?
3. What pieces of data do they have in common in the long term (e.g. after #8206)?

**Comment by kghost** [comment]
> What does `PeerConnectionState` represent?

IMHO, It represents context whose field is only being used by secrue sessions, I'd propose rename it to `SecureSessionContext`.

> we can integrate this state inside PairingSession instead of creating a new class

Totally agree with this apporach.

> What does PairingState represent? What pieces of data do they have in common in the long term (e.g. after Add a peer cache table #8206)?

Only 2 fields will be needed after using the peer cache table.

1. mPeerKeyID
2. mLocalKeyID

`NodeId` is accessible via `PairingSession`, and `PeerAddress` is already exist in the peer table, also accessible via `PairingSession`

**Comment by bzbarsky-apple** [comment]
> It represents context whose field is only being used by secrue sessions

OK, thank you.  Calling it SecureSessionContext makes it sound like it's the object defined in "4.9.2.1. Session Context" in the spec.  Is it?  If not, do we have such an object?  How is it related to what is now called `PeerConnectionState`?

**Comment by kghost** [comment]
> OK, thank you. Calling it SecureSessionContext makes it sound like it's the object defined in "4.9.2.1. Session Context" in the spec. Is it? If not, do we have such an object? How is it related to what is now called `PeerConnectionState`?

`PeerConnectionState` will be renamed to `SecureSessionContext`

> Only two fields are needed in PairingState, or only two fields PairingState and SecureSessionContext/PeerConnectionState have in common?

`PairingState` has no relation with `SecureSessionContext` or `PeerConnectionState`. As PASE and CASE sessions are unsecure, there will be no `SecureSessionContext` associated with them.


Untitled document

**Comment by kghost** [comment]
@bzbarsky-apple please take a look

**Comment by kghost** [comment]
@pan-apple @bzbarsky-apple Please review this PR.

File: src/controller/CHIPDevice.cpp [hunk 4]
```diff
@@ -565,7 +565,7 @@ void Device::OnSessionEstablishmentError(CHIP_ERROR error)
 
 void Device::OnSessionEstablished()
 {
-    mCASESession.PeerConnection().SetPeerNodeId(mDeviceId);
+    mCASESession.SetPeerNodeId(mDeviceId);
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:568]: This is kinda bizarre, in that the session should know which peer we are trying to connect to when we start trying to connect, not when the connection is established!  Pre-existing, but please file a followup to sort out what needs to happen to remove this line.

File: src/controller/CHIPDeviceController.cpp [hunk 5]
```diff
@@ -1182,11 +1182,11 @@ void DeviceCommissioner::OnSessionEstablished()
 
     Device * device = &mActiveDevices[mDeviceBeingPaired];
 
-    mPairingSession.PeerConnection().SetPeerNodeId(device->GetDeviceId());
+    mPairingSession.SetPeerNodeId(device->GetDeviceId());
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1185]: Again, I would have expected the session to know its peer id way before this....

File: src/transport/PairingSession.h [hunk 6]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+    const Transport::PeerAddress & GetPeerAddress() const { return mPeerAddress; }
+    Transport::PeerAddress & GetPeerAddress() { return mPeerAddress; }
+    void SetPeerAddress(const Transport::PeerAddress & address) { mPeerAddress = address; }
+
+    void Clear()
+    {
+        mPeerAddress = Transport::PeerAddress::Uninitialized();
+        mPeerKeyId   = kInvalidKeyId;
+        mLocalKeyId  = kInvalidKeyId;
+    }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:59]: Need to reset mPeerNodeId too?

File: src/transport/PairingSession.h [hunk 7]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
     PairingSession() {}
     virtual ~PairingSession() {}
 
+    NodeId GetPeerNodeId() const { return mPeerNodeId; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:39]: Please document the semantics.  When can this return `kUndefinedNodeId`?  What does that return value mean?

File: src/transport/PairingSession.h [hunk 8]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+    void SetPeerNodeId(NodeId peerNodeId) { mPeerNodeId = peerNodeId; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:40]: Again, feels like this should be known basically at session construction/init time and should not be a public setter API.  Followup for that, please.

File: src/transport/PairingSession.h [hunk 9]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+
+    uint16_t GetPeerKeyId() const { return mPeerKeyId; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:42]: Need to document semantics.  This is not known until the session is far enough advanced, right?  Should this return an `Optional`, perhaps?

File: src/transport/PairingSession.h [hunk 10]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+    void SetPeerKeyId(uint16_t id) { mPeerKeyId = id; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:43]: SetPeerKeyId should not be public API.  Protected, please.

File: src/transport/PairingSession.h [hunk 11]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+    bool IsValidPeerKeyId() const { return mPeerKeyId != kInvalidKeyId; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:44]: Looks unused... But also, if GetPeerKeyId returned `Optional` this is not needed?

File: src/transport/PairingSession.h [hunk 12]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+
+    uint16_t GetLocalKeyId() const { return mLocalKeyId; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:46]: Need to document semantics.  I'd expect this to return a valid value if the session object is in a valid state at all...

File: src/transport/PairingSession.h [hunk 13]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+    void SetLocalKeyId(uint16_t id) { mLocalKeyId = id; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:47]: SetLocalKeyId should not be public API.  Protected, please.

File: src/transport/PairingSession.h [hunk 14]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+    bool IsValidLocalKeyId() const { return mLocalKeyId != kInvalidKeyId; }
+
+    const Transport::PeerAddress & GetPeerAddress() const { return mPeerAddress; }
+    Transport::PeerAddress & GetPeerAddress() { return mPeerAddress; }
+    void SetPeerAddress(const Transport::PeerAddress & address) { mPeerAddress = address; }
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:52]: Should SetPeerAddress() be public API?

File: src/transport/PairingSession.h [hunk 15]
```diff
@@ -36,6 +36,28 @@ class DLL_EXPORT PairingSession
...
+
+    void Clear()
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:54]: Clear should not be public API.  It doesn't make sense to clear the PairingSession but not its subclasses....

**Review by bzbarsky-apple — APPROVED** [review]
Marking approved to unblock this, but please do address the various review comments before resolving them.

File: src/transport/PairingSession.h [hunk 16]
```diff
@@ -77,6 +83,14 @@ class DLL_EXPORT PairingSession
     virtual const char * GetI2RSessionInfo() const = 0;
 
     virtual const char * GetR2ISessionInfo() const = 0;
+
+private:
+    NodeId mPeerNodeId = kUndefinedNodeId;
+    // TODO(#8206): Remove address and use peer cache instead.
+    Transport::PeerAddress mPeerAddress     = Transport::PeerAddress::Uninitialized();
+    static constexpr uint16_t kInvalidKeyId = UINT16_MAX;
```
Comment (bzbarsky-apple) [inline src/transport/PairingSession.h:91]: Also, that's `kAnyKeyId` already.  It's pretty confusing to have different parts of the code attach different meaning to the same value....  Probably better to either use `kAnyKeyId` explicitly as an "invalid" marker or to use a different value and ensure we know what our reserved key id values are...

**Comment by bzbarsky-apple** [comment]
@kghost Note: conflicts.

**Comment by andy31415** [comment]
@kghost this has now all checkmarks, howefver needs conversation resolution

File: src/controller/CHIPDevice.cpp (hunk 4 above)
Comment (kghost) [reply src/controller/CHIPDevice.cpp:583]: Added a TODO tag.

File: src/controller/CHIPDeviceController.cpp (hunk 5 above)
Comment (kghost) [reply src/controller/CHIPDeviceController.cpp:1116]: Added a TODO tag.

File: src/transport/PairingSession.h (hunk 8 above)
Comment (kghost) [reply src/transport/PairingSession.h:43]: Added a TODO tag.

File: src/transport/PairingSession.h (hunk 7 above)
Comment (kghost) [reply src/transport/PairingSession.h:42]: Replaced with optional

File: src/transport/PairingSession.h (hunk 9 above)
Comment (kghost) [reply src/transport/PairingSession.h]: Replaced with optional

File: src/transport/PairingSession.h (hunk 10 above)
Comment (kghost) [reply src/transport/PairingSession.h]: Added a TODO, will be removed

File: src/transport/PairingSession.h (hunk 12 above)
Comment (kghost) [reply src/transport/PairingSession.h:48]: Added a TODO, will be initialized on construction

File: src/transport/PairingSession.h (hunk 14 above)
Comment (kghost) [reply src/transport/PairingSession.h:63]: Added a TODO, will be removed

File: src/transport/PairingSession.h (hunk 15 above)
Comment (kghost) [reply src/transport/PairingSession.h:66]: Added a TODO, should always construct a new instance instead reset the old one

File: src/transport/PairingSession.h (hunk 11 above)
Comment (kghost) [reply src/transport/PairingSession.h]: Replaced with optional

File: src/transport/PairingSession.h (hunk 13 above)
Comment (kghost) [reply src/transport/PairingSession.h:49]: Added a TODO, will be removed

File: src/transport/PairingSession.h (hunk 16 above)
Comment (kghost) [reply src/transport/PairingSession.h:105]: Added a TODO. LocalKeyId must be initialized on construction, PeerKeyId is replaced with an optional.

**Comment by kghost** [comment]
@andy31415  Done, I have already resolved them in the PR.

**Pull request merged**